architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,442 @@
+
+## Objectives
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6688
+
+### Guiding principles
+
+We have two guiding principles for how we make decisions and prioritize investments. 
+- User-defined types must be simple and declarative for application and recipe scenarios.  The highest value scenario is building custom application platforms. We want to build a big ecosystem and so user-defined types should be easy to adopt and easy to develop.
+- User-defined types should be powerful enough for the Radius maintainers to use. We want to leverage user-defined types and declarative functionality as much as possible.
+
+
+
+### Goals
+
+- Support dynamic management of resource types.
+	- eg: Users can register and unregister resource types.
+	- eg: Tooling like the Dashboard and `rad` CLI dynamically query the set of resource types rather than hardcoding a list of known types.
+- Support users defining their own resource types delaratively.
+	- Support rich declarative validation of resource request bodies based on schema.
+	- Support versioning for user-defined resource types.
+		- Including the ability to support multiple API versions.
+		- Including the ability to introduce breaking changes between versions.
+	- eg: Users can define a new resource type and it's schema. Radius will use the schema for validation. Users can add new versions of the resource type, and old ones continue to function.
+- Support use of user-defined types in Bicep with full fidelity.
+	- eg: Users can define their own types and use them in their Bicep application definitions.
+- Support use of user-defined types as part of an application with full fidelity.
+	- User defined types can opt-in to support connections, recipes, application-graph.
+	- eg: Users can define their own types and incorporate them in applications. They work the same as existing types that weren't defined by the user.
+- Support multiple implementation choices for user-defined types.
+	- User-defined types should be easy to implement using declarative functionality.
+	- User-defined types should be powerful enough to support implementing a resource-provider.
+- Cultivate an open-source ecosystem for user-defined types.
+	- Users can find documentation for how to define and publish user-defined types either through open-source or internally to their organization.
+	- Users can find prescriptive guidance to simplify the design of resource schemas. 
+	- Users have a productive inner-loop for developing resource types.
```

Comment:
this would be good to dig into, (perhaps in the detailed docs)- understanding and debugging errors users can encounter during this process.

Reply:
👍 totally agree. I also think this should be covered in the feature spec. /cc @Reshrahim

Reply:
I added some language here to talk about tooling and documentation. I think the follow-up documents are a better place to define this. For now I just want everyone on the same page that it's a goal.